accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+We can now lift the matched value type and the switch case patterns into spaces
+and space unions where our algorithm can operate on them. To determine
+exhaustiveness and reachability from those, we need only one fundamental
+operation, *is-exhaustive*.
 
-     spade  (x: ♣︎, y: ♠︎)  (x: ♦︎, y: ♠︎)  (x: ♥︎, y: ♠︎)  (x: ♠︎, y: ♠︎)
-```
+It takes a space union `value` representing the set of possible values and a set
+of space unions `cases` representing a set of patterns that will match those
+values. It returns `true` if every possible value contained by `value` is
+matched by at least one space in `cases`.
 
-To calculate `(x: Suit, y: Suit) - (x: club, y: spade)`, we need to poke a hole
-in that table:
+Given that operation, we can answer the two exhaustiveness questions like so:
 
-```
-         x: club          diamond       heart         spade
-y:    club  (x: ♣︎, y: ♣︎)  (x: ♦︎, y: ♣︎)  (x: ♥︎, y: ♣︎)  (x: ♠︎, y: ♣︎)
+### Is a switch exhaustive?
 
-   diamond  (x: ♣︎, y: ♦︎)  (x: ♦︎, y: ♦︎)  (x: ♥︎, y: ♦︎)  (x: ♠︎, y: ♦︎)
+To tell if the set of cases in a switch statement or expression are exhaustive
+over the matched value type:
 
-     heart  (x: ♣︎, y: ♥︎)  (x: ♦︎, y: ♥︎)  (x: ♥︎, y: ♥︎)  (x: ♠︎, y: ♥︎)
+1.  Lift the matched value type to a space union `value`.
 
-     spade     <hole>     (x: ♦︎, y: ♠︎)  (x: ♥︎, y: ♠︎)  (x: ♠︎, y: ♠︎)
-```
+2.  Discard any cases that have guards. *Since static analysis can't tell when a
+    guard might evaluate to `false`, any case with a guard doesn't reliably
+    match values and so can't help prove exhaustiveness.*
 
-We don't have a kind of space that naturally represents a "negative" or hole.
-Object spaces model a rectangular region of contiguous cells. They can easily
-represent an entire table, row, column, or cell here. But they can't do a more
-complex shape.
+3.  Lift the remaining case patterns to a set of space unions `cases`.
 
-But we do have unions. We can represent an arbitrarily complex shape using a
-union of simple object shapes that cover everything except the missing parts:
+4.  The switch is exhaustive if is-exhaustive with `value` and `cases` is `true`
+    and `false` otherwise.
 
-```
-         x: club          diamond       heart         spade
-           ┌────────────┐┌────────────────────────────────────────┐
-y:    club │(x: ♣︎, y: ♣︎)││(x: ♦︎, y: ♣︎)  (x: ♥︎, y: ♣︎)  (x: ♠︎, y: ♣︎)│
-           │            ││                                        │
-   diamond │(x: ♣︎, y: ♦︎)││(x: ♦︎, y: ♦︎)  (x: ♥︎, y: ♦︎)  (x: ♠︎, y: ♦︎)│
-           │            ││                                        │
-     heart │(x: ♣︎, y: ♥︎)││(x: ♦︎, y: ♥︎)  (x: ♥︎, y: ♥︎)  (x: ♠︎, y: ♥︎)│
-           └────────────┘│                                        │
-     spade    <hole>     │(x: ♦︎, y: ♠︎)  (x: ♥︎, y: ♠︎)  (x: ♠︎, y: ♠︎)│
-                         └────────────────────────────────────────┘
-```
+### Are any switch cases unreachable?
```

Comment:
To put this section in context we should probably put at the end of the specification more as a side note and mention that this is an unrequired extension of the exhaustiveness checking that can be used by tools to provide hints.

Reply:
I can move this if you want, but I think this is the natural place to stitch together how reachability is defined in terms of the core is-exhaustive function.